v3/src/components/graph/components/chartdots.tsx

```diff
@@ -1,126 +1,157 @@
+    secondaryAxisPlace = attrPlaceToAxisPlace[secondaryAttrPlace] ?? 'left',
+    secondaryAttrID = dataConfig?.attributeID(secondaryAttrPlace),
+    primaryScale = layout.axisScale(primaryAxisPlace) as ScaleBand<string>,
+    secondaryScale = layout.axisScale(secondaryAxisPlace) as ScaleBand<string>
 
   const computeMaxOverAllCells = useCallback(() => {
-    const values = attribute?.strValues,
-      bins: Record<string, number> = {}
-    values?.forEach(aValue => {
-      if (bins[aValue] === undefined) {
-        bins[aValue] = 0
+    const valuePairs = dataConfig?.cases.map(caseID => {
+        return {
+          primary: (primaryAttrID && dataset?.getValue(caseID, primaryAttrID)) ?? '',
+          secondary: (secondaryAttrID && dataset?.getValue(caseID, secondaryAttrID)) ?? '__main__'
+        }
+      }),
+      bins: BinMap = {}
+    valuePairs?.forEach(aValue => {
+      if (bins[aValue.primary] === undefined) {
+        bins[aValue.primary] = {}
+      }
+      if (bins[aValue.primary][aValue.secondary] === undefined) {
+        bins[aValue.primary][aValue.secondary] = 0
       }
-      bins[aValue]++
+      bins[aValue.primary][aValue.secondary]++
     })
-    return Object.keys(bins).reduce((max, aKey) => Math.max(max, bins[aKey]), 0)
-  }, [attribute])
+    return Object.keys(bins).reduce((hMax, hKey) => {
+      return Math.max(hMax, Object.keys(bins[hKey]).reduce((vMax, vKey) => {
+        return Math.max(vMax, bins[hKey][vKey])
+      }, 0))
+    }, 0)
+  }, [dataConfig?.cases, dataset, primaryAttrID, secondaryAttrID])
 
   const refreshPointSelection = useCallback(() => {
-    setPointSelection({dotsRef, dataset, pointRadius: graphModel.getPointRadius(),
-      selectedPointRadius: graphModel.getPointRadius('select')})
+    setPointSelection({
+      dotsRef, dataset, pointRadius: graphModel.getPointRadius(),
+      selectedPointRadius: graphModel.getPointRadius('select')
+    })
   }, [dataset, dotsRef, graphModel])
 
   const refreshPointPositions = useCallback((selectedOnly: boolean) => {
+    // We're pretending that the primaryPlace is the bottom just to help understand the naming
     const
+      primaryCategoriesArray: string[] = dataConfig ? Array.from(dataConfig.categorySetForPlace(primaryAttrPlace)) : [],
+      secondaryCategoriesArray: string[] = dataConfig ?
+        Array.from(dataConfig.categorySetForPlace(secondaryAttrPlace)) : [],
       pointDiameter = 2 * graphModel.getPointRadius(),
       selection = select(dotsRef.current).selectAll(selectedOnly ? '.graph-dot-highlighted' : '.graph-dot'),
       duration = enableAnimation.current ? transitionDuration : 0,
-      cellWidth = primaryScale.bandwidth()
+      primaryCellWidth = primaryScale.bandwidth(),
+      primaryHeight = secondaryScale.bandwidth ? secondaryScale.bandwidth() :
+        layout.axisLength(secondaryAxisPlace),
+      categoriesMap: Record<string, Record<string, { cell: { h: number, v: number }, numSoFar: number }>> = {}
```

Comment:
We haven't talked about this before, and it need not hold up this PR, but it's worth considering whether a JS `Map` would make sense in some of these places where we're using a simple JS object as a map.

Reply:
Interesting. There certainly have been times in V2 when I had to work around the inability to guarantee stable insertion order.